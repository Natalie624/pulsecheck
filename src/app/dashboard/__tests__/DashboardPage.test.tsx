// __tests__/DashboardPage.test.tsx
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import DashboardPage from '../page'
import '@testing-library/jest-dom'
import { mockFetchResponse } from './testUtils/mockFetch'


describe('UI behavior', () => {
  it('renders all UI elements', () => {
    render(<DashboardPage />)

    expect(screen.getByPlaceholderText(/enter your prompt/i)).toBeInTheDocument()
    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /generate/i })).toBeInTheDocument()
  })

  it('updates prompt and tone state', () => {
    render(<DashboardPage />)

    const textarea = screen.getByPlaceholderText(/enter your prompt/i)
    const select = screen.getByRole('combobox')

    fireEvent.change(textarea, { target: { value: 'Weekly update draft' } })
    fireEvent.change(select, { target: { value: 'formal' } })

    expect(textarea).toHaveValue('Weekly update draft')
    expect(select).toHaveValue('formal')
  })

  // EDGE CASE TESTS

  //Empty prompt blocks generation
 it('shows a warning if prompt is empty', async () => {
    render(<DashboardPage />)

    fireEvent.click(screen.getByRole('button', { name: /generate/i }))

    expect(await screen.findByText(/please enter a prompt/i)).toBeInTheDocument()
  })

  // Tone selector has a default
  it('has a default tone selected', () => {
    render(<DashboardPage />)

    const select = screen.getByRole('combobox')
    expect(select).toHaveValue('friendly')
  })

  // Disabled button while loading
  it('disables button while generating', async () => {
    render(<DashboardPage />)

    fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), {
      target: { value: 'Loading test' },
    })

    const button = screen.getByRole('button', { name: /generate/i })
    fireEvent.click(button)

    // Expect disabled state
    expect(button).toBeDisabled()

    // Wait for it to become enabled again (simulate LLM response time)
    await waitFor(() => expect(button).not.toBeDisabled())
  })

  // Charater limit enforced
  it('prevents input over character limit', () => {
    render(<DashboardPage />)

    const textarea = screen.getByPlaceholderText(/enter your prompt/i)
    const longText = 'a'.repeat(2000)

    fireEvent.change(textarea, { target: { value: longText } })

    // TODO: Adjust limit to match your logic, e.g. 1000
    expect(textarea).toHaveValue(longText.slice(0, 1000)) 
  })

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(<DashboardPage />)
    expect(container).toMatchSnapshot()
  })
  
})

// LLM INTEGRATION AND API TESTS

describe('LLM API integration', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  it('renders textarea and button', () => {
    render(<DashboardPage />)
    expect(screen.getByPlaceholderText(/enter your prompt/i)).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /generate/i })).toBeInTheDocument()
  })

  it('shows error if prompt is empty on submit', async () => {
    render(<DashboardPage />)
    fireEvent.click(screen.getByRole('button', { name: /generate/i }))
    expect(await screen.findByText(/please enter a prompt/i)).toBeInTheDocument()
  })

  // TODO: FIX THIS FAILIG TEST OR REWRITE
  it('shows error if prompt exceeds character limit', async () => {
    render(<DashboardPage />)

    const textarea = screen.getByPlaceholderText(/enter your prompt/i)
    fireEvent.change(textarea, { target: { value: 'a'.repeat(2000) } })
    fireEvent.click(screen.getByRole('button', { name: /generate/i }))

   await expect(screen.findByText((text) =>
    text.includes('Prompt exceeds') && text.includes('1000-character limit')
  )).resolves.toBeInTheDocument()
  })

  it('submits prompt and displays summary from API', async () => {
    const mockSummary = 'Here is your summary.'
    mockFetchResponse({ result: { text: mockSummary } })

    render(<DashboardPage />)
    fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), { 
      target: { value: 'We finished Sprint 1.' },
    })

    fireEvent.click(screen.getByRole('button', { name: /generate/i }))

    await waitFor(() => {
      expect(screen.getByText(/generated summary/i)).toBeInTheDocument()
      expect(screen.getByText(mockSummary)).toBeInTheDocument()
    })
  })

  it('shows error if API returns error', async () => {
    mockFetchResponse({ error: 'Bad Request'}, false)

    render(<DashboardPage />)
    fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), {
      target: { value: 'Bad request test' },
    })

    fireEvent.click(screen.getByRole('button', { name: /generate/i }))

    expect(await screen.findByText(/something went wrong/i)).toBeInTheDocument()
  })
  
})
