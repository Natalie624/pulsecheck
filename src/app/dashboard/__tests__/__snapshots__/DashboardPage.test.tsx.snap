// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardPage > matches snapshot 1`] = `
<div>
  <div
    class="max-w-2xl mx-auto py-10 px-4"
  >
    <h1
      class="text-3xl font-bold mb-6"
    >
      PulseCheck
    </h1>
    <textarea
      class="w-full h-32 p-3 border border-gray-300 rounded mb-2"
      placeholder="Enter your prompt here..."
    />
    <div
      class="flex justify-between items-center text-sm mb-2"
    >
      <span
        class="text-gray-500"
      >
        Character limit: 
        0
         / 
        1000
      </span>
    </div>
    <select
      class="w-full p-3 border border-gray-300 rounded mb-4"
    >
      <option
        value="friendly"
      >
        Team Chill
      </option>
      <option
        value="formal"
      >
        Executive Ready
      </option>
      <option
        value="urgent"
      >
        Escalation mode
      </option>
    </select>
    <button
      class="w-full bg-blue-600 text-white py-3 px-4 rounded hover:bg-blue-700 transition disabled:opacity-50"
    >
      Generate
    </button>
  </div>
</div>
`;
